### PR TITLE
Bugfix/14337706/remove state file

### PR DESCRIPTION
**Pull request by ecigar13**
**Reason for Change**:
fix: make user enter a file path to specify state file
fix: auto delete IPAM state to recover when CNI state file is missing and RefCount is not 0.

**Issue Fixed**:

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests

**Notes**:

### CODE REVIEW

File: cni/ipam/ipam.go [hunk 1]
```diff
@@ -167,7 +169,24 @@ func (plugin *ipamPlugin) Add(args *cniSkel.CmdArgs) error {
 
 		// Allocate an address pool.
 		poolID, subnet, err = plugin.am.RequestPool(nwCfg.Ipam.AddrSpace, "", "", options, isIpv6)
+
 		if err != nil {
+			if errors.Is(err, ipam.ErrNoAvailableAddressPools) {
```
Comment (tamilmani1989) [inline cni/ipam/ipam.go:174]: can you move this code to invoker_azure.go?

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_azure.go [hunk 2]
```diff
@@ -25,7 +27,26 @@ type delegatePlugin interface {
 	Errorf(format string, args ...interface{}) *cniTypes.Error
 }
 
+// Create an instance of CNI, then call a delegate function for CNI action (ADD/DEL)
+// Avoid having IPAM state but no CNI state.
 func NewAzureIpamInvoker(plugin *NetPlugin, nwInfo *network.NetworkInfo) *AzureIPAMInvoker {
+	cniStateExists, err := platform.CheckIfFileExists(platform.CNIStateFilePath)
```
Comment (tamilmani1989) [inline cni/network/invoker_azure.go:33]: should do only if it gets no availabe address pool error from ipam

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_azure.go [hunk 3]
```diff
@@ -57,20 +57,20 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
 	if err != nil {
 
 		if errors.Is(err, ipam.ErrNoAvailableAddressPools) {
-			cniStateExists, err := platform.CheckIfFileExists(platform.CNIStateFilePath)
+			cniStateExists, er := platform.CheckIfFileExists(platform.CNIStateFilePath)
```
Comment (rbtr) [inline cni/network/invoker_azure.go:60]: why?

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_azure.go (hunk 3 above)
Comment (ecigar13) [reply cni/network/invoker_azure.go]: I fixed it.

File: cni/network/invoker_azure.go [hunk 4]
```diff
@@ -97,6 +122,24 @@ func (invoker *AzureIPAMInvoker) Delete(address *net.IPNet, nwCfg *cni.NetworkCo
 
 	if address == nil {
 		if err := invoker.plugin.DelegateDel(nwCfg.Ipam.Type, nwCfg); err != nil {
+			if errors.Is(err, ipam.ErrNoAvailableAddressPools) {
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply cni/network/invoker_azure.go:126]: can you remove this from delete call?

File: cni/network/invoker_azure.go [hunk 5]
```diff
@@ -48,7 +53,28 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
...
+		}
+
+		if !cniStateExists && ipamStateExists {
+			er = os.Remove(platform.CNIIpamStatePath)
+			if er != nil {
+				log.Printf("[cni] error deleting state file %v", er)
+			}
+		}
+
+		log.Printf("[cni] Deleted IPAM state file")
```
Comment (tamilmani1989) [inline cni/network/invoker_azure.go:74]: can you move this code to separate function?

File: cni/network/invoker_azure_test.go [hunk 6]
```diff
@@ -132,12 +133,13 @@ func TestAzureIPAMInvoker_Add(t *testing.T) {
...
-		args    args
-		want    *cniTypesCurr.Result
-		want1   *cniTypesCurr.Result
-		wantErr bool
+		name        string
+		fields      fields
+		args        args
+		want        *cniTypesCurr.Result
+		want1       *cniTypesCurr.Result
+		wantErrType error
```
Comment (tamilmani1989) [inline cni/network/invoker_azure_test.go:141]: why did you add wantErrType..?

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_azure.go [hunk 7]
```diff
@@ -112,6 +95,26 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
 	return addResult, err
 }
 
+func (invoker *AzureIPAMInvoker) deleteIpamState() {
+	log.Printf("[cni] Deleting IPAM state file")
```
Comment (tamilmani1989) [inline cni/network/invoker_azure.go:99]: move this log to line 111 before os.remove call

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_azure_test.go (hunk 6 above)
Comment (tamilmani1989) [reply cni/network/invoker_azure_test.go:141]: revert this change if not used

**Review by tamilmani1989 — APPROVED** [review]
lgtm 🚢

File: cni/network/invoker_azure.go [hunk 8]
```diff
@@ -48,6 +53,10 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
 
 	// Call into IPAM plugin to allocate an address pool for the network.
 	addResult.ipv4Result, err = invoker.plugin.DelegateAdd(addConfig.nwCfg.Ipam.Type, addConfig.nwCfg)
+
+	if err != nil && strings.Contains(err.Error(), ipam.ErrNoAvailableAddressPools.Error()) {
```
Comment (rbtr) [inline cni/network/invoker_azure.go:57]: is there an issue with doing this via `errors.Is/As`?

File: ipam/api.go [hunk 9]
```diff
@@ -18,7 +18,7 @@ var (
 	errAddressPoolNotFound     = fmt.Errorf("Address pool not found")
 	errAddressPoolInUse        = fmt.Errorf("Address pool already in use")
 	errAddressPoolNotInUse     = fmt.Errorf("Address pool not in use")
-	errNoAvailableAddressPools = fmt.Errorf("No available address pools")
+	ErrNoAvailableAddressPools = fmt.Errorf("No available address pools")
```
Comment (rbtr) [inline ipam/api.go:21]: it's easier to find them when you put exported vars in a separate block

Comment (rbtr) [reply ipam/api.go:21]: because they might be wrapped or appear in logs, error strings should not be sentence case (and should not have closing punctuation)

File: store/json.go [hunk 10]
```diff
@@ -42,9 +39,8 @@ type jsonFileStore struct {
 // NewJsonFileStore creates a new jsonFileStore object, accessed as a KeyValueStore.
 func NewJsonFileStore(fileName string, lockclient processlock.Interface) (KeyValueStore, error) {
 	if fileName == "" {
-		fileName = defaultFileName
+		return &jsonFileStore{}, errors.New("Need to pass in a json file path")
```
Comment (rbtr) [inline store/json.go:42]: no sentence case

File: cni/network/invoker_azure.go [hunk 11]
```diff
@@ -86,6 +95,29 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
...
+		return
+	}
+
+	ipamStateExists, err := platform.CheckIfFileExists(platform.CNIIpamStatePath)
+	if err != nil {
+		log.Printf("[cni] Error checking IPAM state exist: %v", err)
+		return
+	}
+
+	if !cniStateExists && ipamStateExists {
```
Comment (rbtr) [inline cni/network/invoker_azure.go:111]: could drop this and return immediately if `cniStateExists` above?

Comment (rbtr) [reply cni/network/invoker_azure.go:111]: similarly could you return immediately if `!ipamStateExists`

Comment (tamilmani1989) [reply cni/network/invoker_azure.go:111]: agree on 1st case..2nd case would not happen

File: cni/network/invoker_azure.go (hunk 8 above)
Comment (tamilmani1989) [reply cni/network/invoker_azure.go:57]: yes this error is from ipam and the ipam code unwrapping to string before writing to stdout and in cni side Error.Is couldn't find it

Comment (ecigar13) [reply cni/network/invoker_azure.go:57]: From CNI specification: https://www.cni.dev/docs/spec/#Error
CNI delegate actions will type erase the error to a JSON string, so `errors.Is/As` won't work. Only way to check error type is doing it inside IPAM. Outside delegated functions, I can use string contains, or parse into a JSON and also use string contains for the Msg field.

File: ipam/api.go (hunk 9 above)
Comment (ecigar13) [reply ipam/api.go:21]: Do I change all other error strings to lower case, too? They're not part of this work item: 14337706.

Comment (rbtr) [reply ipam/api.go:21]: 😕
You did the right thing and matched the convention here. But it could be something to file as a future repair, and keep in mind when creating new error strings later.
